docs/manual/src/udl/delegates.md

```diff
@@ -0,0 +1,242 @@
+        }
+    }
+
+    fun <T> withErrorReporter(rustCall: () -> T) =
+        try {
+            rustCall()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+
+    fun thenNotifyListeners(obj: DemoLib, () -> T) {
+        try {
+            rustCall()
+        } finally {
+            notifyListeners(obj)
+        }
+    }
+}
+```
+
+Then, we can re-write the `DemoLib` as:
+
+```kotlin
+class DemoLib(
+    val delegate: DemoDelegate = MyDemoDelegate()
+) {
+    val demoRustObject = new DemoRustObject()
+
+    fun doExpensiveThing() = delegate.onBackgroundThread {
+        demoRustObject.doExpensiveThing()
+    }
+
+    fun doCrashyThing() = delegate.withErrorReporter {
+        demoRustObject.doCrashyThing()
+    }
+
+    fun getStateAt(index: Int) = delegate.withErrorReporter {
+        rust.getStateAt(index)
+    }
+
+    fun setStateAt(index: Int, value: String) = delegate.thenNotifyListeners(this) {
+        rust.setStateAt(index, value)
+    }
+}
+```
+
+This is much better, but it's still looking a bit cut and pasty. Enter delegate objects.
+
+## Solution
+
+```webidl
+namespace demo {}
+
+[Delegate]
+interface DemoDelegate {
+    any  with_error_reporter();
+    void on_background_thread();
+    any then_notify_listeners();
+}
+
+[Delegate=DemoDelegate]
+interface DemoRustObject {
+    [CallWith=on_backghround_thread]
+    void do_expensive_thing();
+
+    [Throws=RustyError, CallWith=with_error_reporter]
+    void do_crashy_thing();
+
+    [CallWith=with_error_reporter]
+    string get_state_at(i32 index);
+
+    [CallWith=then_notify_listeners]
+    void set_state_at(i32 index, string value);
+}
+```
+
+With this UDL: the `[Delegate]` annotation declares an `DemoDelegate` as a delegate object. Delegate objects never cross the FFI.
+
+They are implemented as a `protocol` in Swift, and `interface` in Kotlin.
+
+They declare zero argument methods, but can return an arbitrary concrete type, `void` or the generic `any` type. The `DemoDelegate` interface is now generated,
+and each method accepts the delegating object and a generic closure which will call the Rust code.
+
+```kotlin
+// generated by uniffi
+interface DemoDelegate<ObjectType> {
+    fun <ReturnType> onBackgroundThread(obj: ObjectType, rustCall: () -> ReturnType)
+    fun <ReturnType> withErrorReporter(obj: ObjectType, rustCall: () -> ReturnType): ReturnType
+    fun <ReturnType> thenNotifyListeners(obj: ObjectType, rustCall: () -> ReturnType): ReturnType
+}
+```
+
+The `[Delegate=DemoDelegate]` annotation above the `DemoRustObject` `interface` declaration ties the delegate object to our original `DemoRustObject` class.
+
+This changes the generated API of `DemoRustObject`:
+
+* it adds a `DemoDelegate` argument to every constructor.
+* it changes the return types and throws types for each method called with a delegate method to match the delegate method. Where the delegate method returns `any`, the original return type remains.
```

Comment:
My initial reaction to these changes is that they seem too magical, especially the extra argument to each constructor. You can't see this extra argument by looking at the Rust code, or by looking at the UDL for the construtors, you just have to know that this is what happens when you use the special `[Delegate]` annotation.

What would it look like to specify the delegate argument explicitly in each constructor definition in the UDL? Would it be worth it, to have the UDL more directly match the shape of the generated code?

Reply:
Yes, it does feel magical.

Other ways we could do: pass in one or more decorators into each constructor. Some amount of error checking would be needed to ensure that every constructor had the same decorators. This would allow the user to specify where in the arg list the decorator objects came.